### ISSUE TITLE
Different accuracy for quantized model when inferring from one image vs batch

### ISSUE THREAD

**Issue by dhingratul** [tensorflow/tensorflow#20602]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: ubuntu 16.04
- **TensorFlow installed from (source or binary)**: pip
- **TensorFlow version (use command below)**: 1.8
- **Python version**: 2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**: gtx 1080ti
- **Exact command to reproduce**:

```
!python quantize_graph.py \
--input=logs_vgg/frozen_model.pb \
--output_node_names="vgg_16/fc8/squeezed" --print_nodes --output=logs_vgg/quantized_graph.pb \
--mode=eightbit --logtostderr
```
```
tf.reset_default_graph()
tf.set_random_seed(0)
g2 = load_graph('logs_vgg/quantized_model.pb')
ctr = 0
with g2.as_default():
    with tf.Session(graph=g2) as sess:
        restored_x = g2.get_tensor_by_name('import/Inputs:0')
        restored_logits = g2.get_tensor_by_name('import/vgg_16/fc8/squeezed:0')
        times = []
        for i in range(num_images):
            t0 = time.time()
            a = sess.run(restored_logits, feed_dict={restored_x:np.expand_dims(images[i, :, :, :], axis=0)})
            t1 = time.time()
            total_time = t1 - t0
            times.append(total_time)
            pred = np.argsort(a[0])
            pred = pred[-5:]
            if y[i] in set(pred):
                ctr += 1   
print("TF Prediction acc:" + str(ctr/float(num_images)))
print("Inference time(ms): " + str((sum(times[1:])/float(num_images - 1)) * 1000))
```
TF Prediction acc:0.79
Inference time(ms): 56.0006372856

```
tf.reset_default_graph()
tf.set_random_seed(0)
g = load_graph('logs_vgg/quantized_model.pb')
with g.as_default():
    with tf.Session(graph=g) as sess:
        restored_x = g.get_tensor_by_name('import/Inputs:0')
        restored_logits = g.get_tensor_by_name('import/vgg_16/fc8/squeezed:0')
        t0 = time.time()
        a = sess.run(restored_logits, feed_dict={restored_x:images})
        t1 = time.time()
        total_time = t1 - t0
ctr = 0
for i in range(num_images):
    pred = np.argsort(a[i])
    pred = pred[-5:]
    if y[i] in set(pred):
        ctr += 1   
print("TF Prediction acc:" + str(ctr/float(num_images)))
print("Inference time(ms): " + str(total_time/float(num_images) * 1000))
```

 TF Prediction acc:0.76
Inference time(ms): 406.530380249

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

Different accuracy for quantized model when inferring from one image vs batch

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by suharshs** [comment]
Are you sure you are comparing the exact same images? If so, you should be able to find a single image that gets a different prediction in single image and batch.

**Comment by dhingratul** [comment]
Yes, the images are the same.

**Comment by dhingratul** [comment]
This only happens with a quantized model, for a regular model, this is not reproducible.

**Comment by suharshs** [comment]
This is because the quantize_graph does dynamic quantization per input tensor. So if the batch is larger then the min max will be worse for accuracy. This is working as intending as quantization efforts have been focused to TFLite. https://www.tensorflow.org/performance/quantization

We will also be offering some easier tools for quantization soon that should help with this.

**Comment by dhingratul** [comment]
Is there any efforts for using tf.lite on GPU ?
